modules/features2d/test/test_descriptors_invariance.impl.hpp

```diff
@@ -12,6 +12,19 @@ typedef tuple<std::string, Ptr<FeatureDetector>, Ptr<DescriptorExtractor>, float
     String_FeatureDetector_DescriptorExtractor_Float_t;
 
 
+static void removeVerySmallKeypoints(vector<KeyPoint>& keypoints)
```

Comment:
I don't like the idea of unconditionally removing -1 octave since the octave was added to evaluation intentionally and extends the amount of useful keypoints significantly in accordance with the paper. However it make sense to filter out points of low quality. I've looked through the code and found out that the ContrastThreshold is set to 0.04 while the paper states 0.03. However as far as I could understand this value was used as is in the paper but scaled down in nLayer times in the [implementation](https://github.com/opencv/opencv/blob/6fbf7042b5daf0dd0d5ad553ba266de0a497c0ba/modules/features2d/src/sift.simd.hpp#L381)

I've tried to change threshold to 0.09 to conform with the paper and it fixed the DetectorInvariance test

Reply:
That is strange. So should I change it to `if( std::abs( contr ) < contrastThreshold )` and set `contrastThreshold` to 0.03?

Reply:
I think the code should be retained as is. IMO it would be better to notice the fact in documentation and increase the threshold (implicitly by changing the default value or explicitly (preferred option) stating values for SIFT constructor in the test)

Reply:
I removed removeVerySmallKeypoints from both test impls, but in my environment the DetectorInvariannce test wasn't passed even if ContrastThreshold is 0.09.
```
[----------] 1 test from SIFT/DetectorScaleInvariance
[ RUN      ] SIFT/DetectorScaleInvariance.scale/0, where GetParam() = ("detectors_descriptors_evaluation/images_datasets/bikes/img1.png", 0.69, 0.98)
scale = 1.5, keyPointMatchesRatio = 0.767554, scaleInliersRatio = 0.996845
scale = 2, keyPointMatchesRatio = 0.812734, scaleInliersRatio = 1
/home/yosshi999_for_entry/opencv/modules/features2d/test/test_detectors_invariance.impl.hpp:198: Failure
Expected: (keyPointMatchesRatio) >= (minKeyPointMatchesRatio), actual: 0.675556 vs 0.69
scale = 2.5, keyPointMatchesRatio = 0.675556, scaleInliersRatio = 1
```

Reply:
I've reviewed changes applied during investigation and noticed the change in test implementation as well. I've removed [these ](https://github.com/terfendail/opencv/blob/fdb8b3dfa887738f48d742f451b01391066ea2d9/modules/features2d/test/test_detectors_invariance.impl.hpp#L36) two lines. Such filter for keypoint matching could cause matching to the first encountered instead of best match if close enough keypoint with no good pair was checked prior(which is quite often since keypoint vector starts with a crowd of low quality keypoints from the -1 octave).

Reply:
I confirmed that the test is passed, but in that case some match pairs sometimes have same point. I'm reading the usage and source code of BFMatcher in order to make that implementation more reasonable.

Reply:
I just removed these lines because I couldn't find computationally efficient solution.